### ISSUE TITLE
z coordinate in Pose tracking

### ISSUE THREAD

**Issue by RaiShra1** [google/mediapipe#1611]
Requesting info about the integration of z coordinate / depth info on the landmarks of the pose tracker.

**Comment by rmothukuru** [comment]
@RaiShra1,
Can you please elaborate? Thanks!

**Comment by RaiShra1** [comment]
Currently in the pose tracking, z coordinate of the landmarks do not constitute for the depth of the subject. Depth information is a critical point for my research. So wanted to know if google would release the models with depth information from the landmarks in the future?

**Comment by M00NSH0T** [comment]
Just adding my two cents here, I agree it would be great to get an update on where things stand with the z-coordinate.


Citing the [documentation](https://google.github.io/mediapipe/solutions/pose_classification.html#future-work):
```Future Work
We are actively working on improving BlazePose GHUM 3D’s Z prediction. It will allow us to use joint angles in the feature vectors, which are more natural and easier to configure (although distances can still be useful to detect touches between body parts) and to perform rotation normalization of poses and reduce the number of camera angles required for accurate k-NN classification.```

The documentation is also a bit inconsistent with what we should / shouldn't be doing with the z coordinate currently. [Here](https://developers.google.com/ml-kit/vision/pose-detection/classifying-poses) it says we could try extending angle calculation to 3 dimensions to see if it performs better (which I'm currently coding in to my application to test), but the z-coordinate [blurb ](https://developers.google.com/ml-kit/vision/pose-detection#z_coordinate) says this isn't a true 3D coordinate, so I have my doubts about the results I'll see. Probably better to replace the KNN and surrounding logic with a neural network or random forest classifier at the same time in case there's some strange non-linear scaling going on with it. Along those lines, the extended notebook example only uses the X and Y coordinate to normalize scale, and in the KNN distance calculations are a bit strange in that they're not actually calculating euclidiean distance, but the mean and max of the distances in each coordinate direction. And then, the distance in the z-direction is weighted at 0.2 of what X and Y get. Is that because the scale of this axis is 5x greater (on average) or that it's just so unreliable that you basically want to ignore it? If we can have any reasonable degree of confidence in Z, we should be able to just use np.linalg.norm to get the distance between points and then have a much simple embedding than this 3D one? I figure while I'm adding in the 3D angles, I may as well also try that, but I really wish there was a bit more info on the dependability / scale / upcoming changes to Z, especially since it's been almost a year since this all came out.

Where I'm coming from here is that I've been trying to get this to work for a more complex (martial arts) application, but detecting poses using the current embeddings is highly inaccurate. Without a dependable z-coordinate, we effectively need at least dozens (more like hundreds) of sample images from dozens of camera angles for each pose class to get this to work. And I have dozens of classes in my application, so I've spent hundreds of hours at this point filming and labeling my current dataset of 50k images, but the output is just all over the place. I'm playing whack-a-mole with my classes... getting more and more images from more and more angles for problem classes, filtering outliers, etc, but in the end I fear that it's going to just be wasted effort if all of a sudden you provide us with a more accurate z-coordinate since that will make the embedding much more generalizable and cut down on the need for all these different camera angles for each class. It'd be better for me to be focusing on capturing more of the nuance of the moves / poses.

Anyway, any update would be super appreciated.

**Comment by sgowroji** [comment]
Hi @RaiShra1 @M00NSH0T, Thanks for sharing your concern.  We have added the right owners who can resolve your query.

**Comment by Cris21395** [comment]
Any update on this?

Thank you.

**Comment by Custards1** [comment]
Has any progress been made on this?

**Comment by ivan-grishchenko** [comment]
Hi. Recently we updated the released model to produced additional world 3D landmarks output with origin in hips center.
More details can be found in API: https://google.github.io/mediapipe/solutions/pose#pose_world_landmarks
And in the blog: https://blog.tensorflow.org/2021/08/3d-pose-detection-with-mediapipe-blazepose-ghum-tfjs.html

If you want to calculate relative depth between landmarks, this output should be enough for you. But if you want to calculate distance to the camera, extra work needs to be done. Having both 2D and 3D landmarks and knowing camera parameters you can apply some optimization (e.g. Procrustes analysis) to determine translation in 3D space.
